File: arrays_and_strings/URLify.cpp
// Write a method to replace all spaces in a string with '%20'. You may assume that
// the string has suffice space at the end to hold the additional characters, and
// that you are given the "true" length of the string.
#include <iostream>
#include <string>
using namespace std;

string URLify(string s, int trueLen) {
    int spaceCount = 0;
    for (int i = 0; i < trueLen; i++) {
        if (s[i] == ' ') spaceCount++;
    }
    int idx = trueLen + spaceCount * 2;
    int len = s.length();
    if (trueLen <  len) s[trueLen] = '\0'; // end array
    for (int i = trueLen - 1; i >= 0; i--) {
        if (s[i] == ' ') {
            s[idx - 1] = '0';
            s[idx - 2] = '2';
            s[idx - 3] = '%';
            idx = idx - 3;
        } else {
            s[idx - 1] = s[i];
            idx--;
        }
    }

    return s;
}

int main () {
    string s = "Mr John Smith    ";
    int str_length = 13;

    cout << URLify(s, str_length);

    cout << endl;
    return 0;
}

File: arrays_and_strings/urlify.py
"""write a method to repace all spaces in a string with '%20'. You may assume
that the string has sufficient space at the end to hold the additional characters,
and that you are given the "true" length of the string.
"""
import unittest

def urlify(string, length):
    """function replaces single space with %20 and removes trailing spaces"""
    new_idx = len(string)

    for i in reversed(range(length)):
        if string[i] == ' ':
            # replace spaces
            string[new_idx - 3: new_idx] = '%20'
            new_idx -= 3
        else:
            # move characters
            string[new_idx - 1] = string[i]
            new_idx -= 1

    return string


class Test(unittest.TestCase):
    """docstring for Test."""
    # using lists because python strings are immutable
    data = [
        (list('Mr John Smith    '), 13, list('Mr%20John%20Smith')),
        (list('hello world  '), 11, list('hello%20world'))
    ]

    def test_urlify(self):
        """testing"""
        for [test_string, length, expected] in self.data:
            actual = urlify(test_string, length)
            self.assertEqual(actual, expected)

if __name__ == '__main__':
    unittest.main()
